src/wallet/wallet.cpp

```diff
@@ -3862,41 +3860,49 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
     }
 
     CAmount nMinFee = 0;
-    CAmount nDeduct = 0;
-    bool bOutputsComplete = false;
+    int maxOutputs = 30;
     CAmount nMinFeeBase = (IsProtocolV07(txNew.nTime) ? MIN_TX_FEE : MIN_TX_FEE_PREV7);
     while(true)
     {
+        // Clear outputs
+        txNew.vout.erase(txNew.vout.begin() + 1u+bMinterKey, txNew.vout.end());
+
+        // Assume success
+        bool outputsOk = true;
+
         // split and set amounts based on rfc28
-        if (!bOutputsComplete) {
-            if (pwallet->m_split_coins) {
-                CAmount current = nCredit - nMinFee;
-                double ratio = current / nTarget;
-                int desiredOutputs = int(std::floor((std::sqrt(4 * std::pow(ratio, 2) + 1) + 1) / 2));
-                if (bDebug)
-                    LogPrintf("rfc28: security level is %f, target amount %f, desired outputs %d from %f ppc\n",
-                        securityLevel, double(nTarget)/COIN, desiredOutputs, double(current)/COIN);
-                CAmount remaining = current;
-                for (int i=1; i<desiredOutputs; i++) {
-                    txNew.vout.push_back(CTxOut(nTarget, scriptPubKeyOut));
-                    remaining -= nTarget;
-                    // make sure transaction below 1kb
-                    if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 950)
-                        break;
+        if (pwallet->m_split_coins) {
+            CAmount current = nCredit - nMinFee;
+            double ratio = current / nTargetOutputAmount;
+            // Obtain the optimal number of outputs and clamp it to maxOutputs to ensure the fee is not exceeded
+            int desiredOutputs = std::min(
+                int(std::floor((std::sqrt(4 * std::pow(ratio, 2) + 1) + 1) / 2)),
+                maxOutputs
+            );
+
+            if (bDebug)
+                LogPrintf("rfc28: security level is %f, target amount %f, desired outputs %d from %f ppc\n",
+                    securityLevel, double(nTargetOutputAmount)/COIN, desiredOutputs, double(current)/COIN);
+
+            CAmount outValue = current / desiredOutputs;
+            CAmount remainder = current - outValue*desiredOutputs;
+
+            for (int i=0; i<desiredOutputs; i++) {
+                // Add remainder to first output
+                txNew.vout.push_back(CTxOut(outValue + (i == 0 ? remainder : 0), scriptPubKeyOut));
+                // make sure transaction below 1kb
+                if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 1024) {
```

Comment:
Isn't the fee-free amount 1000 bytes?

Reply:
Also I think we should add an extra condition that `i` must be greater than 0. That shouldn't happen but it protects against the possibility that `maxOutputs` would get set to 0.